### PR TITLE
OCSP Stapling Enhancement for Apache

### PR DESCRIPTION
**Pull request by sagi**
- Currently supports only ~~2.4~~ >=2.3.3.
- `letsencrypt --staple-ocsp -d dumpbits.com` [no problem to set it on for apache => 2.3.3]
-  To check OCSP Stapling:
~~~
[~]$ echo QUIT | openssl s_client -connect dumpbits.com:443 -status 2>/dev/null | grep -A 31 'OCSP Resp'
OCSP Response Data:
    OCSP Response Status: successful (0x0)
    Response Type: Basic OCSP Response
    Version: 1 (0x0)
    Responder Id: C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
    Produced At: Mar 26 17:54:00 2016 GMT
    Responses:
    Certificate ID:
      Hash Algorithm: sha1
      Issuer Name Hash: 7EE66AE7729AB3FCF8A220646C16A12D6071085D
      Issuer Key Hash: A84A6A63047DDDBAE6D139B7A64565EFF3A8ECA1
      Serial Number: 032A2108AAA650E6EE2E6B041C03C2612A19
    Cert Status: good
    This Update: Mar 26 17:00:00 2016 GMT
    Next Update: Apr  2 17:00:00 2016 GMT

    Signature Algorithm: sha256WithRSAEncryption
         64:f2:71:02:6a:97:d9:eb:13:c1:5c:7a:f5:eb:26:89:3b:40:
         e3:08:82:f7:71:d4:fa:61:4a:8e:4a:7d:e9:53:84:e9:3a:89:
         67:66:08:d9:0e:79:65:9a:8d:dc:fb:07:cc:93:4f:eb:4e:3c:
         cc:7f:cd:fd:db:8f:c3:25:c3:54:87:a9:9c:35:6f:c1:39:31:
         e0:b1:f6:b1:3d:52:5d:db:bb:69:0f:23:05:fe:33:29:1f:ff:
         c6:af:17:a5:98:58:50:3a:48:93:5c:09:4b:f3:91:36:48:31:
         ed:ee:47:4d:66:c3:25:cf:56:b7:f4:48:80:eb:b8:f0:27:b1:
         97:18:b4:88:71:c6:55:5d:bb:25:16:48:98:85:8a:12:8d:64:
         bf:51:df:39:b1:44:91:e1:f2:c6:c3:7d:23:2b:d2:0f:4c:7f:
         57:b1:c9:ae:ec:32:b5:6a:87:bd:83:43:f1:f7:3c:8c:11:5c:
         9d:a5:12:fa:e6:79:87:45:c6:1d:46:c8:14:1e:8d:d1:de:7a:
         0d:e4:53:f2:c9:b6:e5:6e:cb:91:14:bb:04:38:36:4f:71:55:
         e1:ff:71:c7:a6:31:ed:db:6c:0f:d7:f5:ef:0c:6e:08:6b:e0:
         37:cf:ca:a5:67:89:c2:de:8e:36:6d:2f:41:7f:9f:10:c6:de:
         4d:b1:2d:09
======================================

~~~

### CODE REVIEW

**Comment by sagi** [comment]
Implements #930.

**Comment by pde** [comment]
@joohoi are you interested in reviewing?

File: letsencrypt-apache/letsencrypt_apache/configurator.py [hunk 1]
```diff
@@ -966,6 +967,66 @@ def enhance(self, domain, enhancement, options=None):
...
+
+            ssl_vhost_aug_path = parser.get_aug_path(ssl_vhost.filep)
+
+            # Check if there's an existing SSLStaplingCache directive.
+            stapling_cache_aug_path = self.parser.find_dir('SSLStaplingCache',
+                    None, ssl_vhost_aug_path)
+
+            # We'll simply delete the directive, as it might be something like
+            # SSLStaplingCache /tmp/ocsp_stapling
+            # The OS usually cleans /tmp on reboot, which means that the server
```
Comment (jsha) [inline letsencrypt-apache/letsencrypt_apache/configurator.py]: Note that this is not really a big problem - if a server re-requests the OCSP response on reboot, it's not a big deal. Still, it's nice to have a consistent cache path.

File: letsencrypt-apache/letsencrypt_apache/configurator.py [hunk 2]
```diff
@@ -966,6 +967,66 @@ def enhance(self, domain, enhancement, options=None):
...
+            # Would have to query LetsEncrypt's OCSP Responder (akamai) again.
```
Comment (jsha) [inline letsencrypt-apache/letsencrypt_apache/configurator.py]: The fact that Akamai is the CDN for our OCSP responder is probably too much detail to include in this particular comment.

File: letsencrypt-apache/letsencrypt_apache/configurator.py [hunk 3]
```diff
@@ -966,6 +967,66 @@ def enhance(self, domain, enhancement, options=None):
...
+            # Then we'll plant a new directive
+            if stapling_cache_aug_path:
+                self.aug.remove(
+                        re.sub(r"/\w*$", "", stapling_cache_aug_path[0]))
+
+            self.parser.add_dir_to_ifmodssl(ssl_vhost_aug_path,
+                    "SSLStaplingCache",
+                    ["shmcb:/var/run/apache2/stapling_cache(128000)"])
+
+            msg = "OCSP Stapling was enabled to SSL Vhost: %s.\n"%(
```
Comment (jsha) [inline letsencrypt-apache/letsencrypt_apache/configurator.py]: to -> on

File: letsencrypt/cli.py [hunk 4]
```diff
@@ -674,6 +675,17 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):
         " upgrade-insecure-requests\" header to every HTTP response.",
         dest="uir", default=None)
     helpful.add(
+        "security", "--staple-ocsp", action="store_true",
+        help=" Enables OCSP Stapling. A valid OCSP response is stapled to"
```
Comment (jsha) [inline letsencrypt/cli.py]: There should be no space between `"` and `Enables.

File: letsencrypt/cli.py [hunk 5]
```diff
@@ -674,6 +675,17 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):
...
+        " the certificate that the server offers during TLS.",
+        dest="staple", default=None)
+    helpful.add(
+        "security", "--no-staple-ocsp", action="store_false",
+        help=" Do not automatically enable OCSP Stapling.",
```
Comment (jsha) [inline letsencrypt/cli.py]: There should be no space between `"` and `Do`.

**Comment by joohoi** [comment]
I'd like to see an actual vhost file in Apache testdata with stapling directives already in place. We're also handling the enabling of modules only for Debian & Ubuntu currently, so I think it'd be good to notify users about the need to enable socache_shmcb module (which I don't think we have actually required in the past),  if the apache reload check after deployment fails.

**Comment by jsha** [comment]
@sagi: Any news on this?

**Comment by sagi** [comment]
Still working on joohoi's comments.

**Comment by sagi** [comment]
Sorry it has taken that much time. I'm almost done, merging rename related stuff.

**Comment by sagi** [comment]
I'll try to add a notification regarding the socache_shmcb module on failed reboot tomorrow.

**Comment by jsha** [comment]
Is this ready for review?

File: certbot-apache/certbot_apache/parser.py [hunk 6]
```diff
@@ -200,6 +200,7 @@ def add_dir_to_ifmodssl(self, aug_conf_path, directive, args):
         # TODO: Add error checking code... does the path given even exist?
         #       Does it throw exceptions?
         if_mod_path = self._get_ifmod(aug_conf_path, "mod_ssl.c")
+
```
Comment (jsha) [inline certbot-apache/certbot_apache/parser.py]: This is a spurious diff, please remove.

File: certbot-apache/certbot_apache/obj.py [hunk 7]
```diff
@@ -164,7 +164,6 @@ def __eq__(self, other):
                     self.ssl == other.ssl and
                     self.enabled == other.enabled and
                     self.modmacro == other.modmacro)
-
```
Comment (jsha) [inline certbot-apache/certbot_apache/obj.py]: Spurius diff, please restore.

**Comment by pde** [comment]
Looks like this won't land today, so kicking to 0.7.0; we can probably finish this while Sagi is in town...

**Comment by sagi** [comment]
If a `mods-enabled / mods-available layout` isn't detected (i.e. its not a Debian-like host) the `enable_mod` function notifies the user of to manually enable the module.

**Comment by sagi** [comment]
I think this is ready for a review.

File: certbot-apache/certbot_apache/configurator.py [hunk 8]
```diff
@@ -373,6 +374,7 @@ def _find_best_vhost(self, target_name):
         # Points 3 - Servername no SSL
         # Points 2 - Wildcard no SSL
         # Points 1 - Address name with no SSL
+
```
Comment (jsha) [inline certbot-apache/certbot_apache/configurator.py]: Remove this spurious diff.

File: certbot-apache/certbot_apache/configurator.py [hunk 9]
```diff
@@ -594,7 +596,6 @@ def add_name_vhost(self, addr):
 
         """
         loc = parser.get_aug_path(self.parser.loc["name"])
-
```
Comment (jsha) [inline certbot-apache/certbot_apache/configurator.py]: Remove this spurious diff.

File: certbot-apache/certbot_apache/configurator.py [hunk 10]
```diff
@@ -971,6 +972,72 @@ def enhance(self, domain, enhancement, options=None):
...
+            if not use_stapling_aug_path:
+                self.parser.add_dir(ssl_vhost.path, "SSLUseStapling", "on")
+
+            ssl_vhost_aug_path = parser.get_aug_path(ssl_vhost.filep)
+
+            # Check if there's an existing SSLStaplingCache directive.
+            stapling_cache_aug_path = self.parser.find_dir('SSLStaplingCache',
+                    None, ssl_vhost_aug_path)
+
+            # We'll simply delete the directive, as it might be something like
```
Comment (jsha) [inline certbot-apache/certbot_apache/configurator.py]: Remove this comment

File: certbot-apache/certbot_apache/configurator.py [hunk 11]
```diff
@@ -971,6 +972,72 @@ def enhance(self, domain, enhancement, options=None):
             logger.warn("Failed %s for %s", enhancement, domain)
             raise
 
+    def _enable_ocsp_stapling(self, ssl_vhost, unused_options):
+        """Enables OCSP Stapling
```
Comment (jsha) [inline certbot-apache/certbot_apache/configurator.py]: Add a comment that this implicitly depends on SSLCertificateChainFile being set by other code.

File: certbot/cli.py [hunk 12]
```diff
@@ -704,6 +704,7 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):
     helpful.add(
         "security", "--rsa-key-size", type=int, metavar="N",
         default=flag_default("rsa_key_size"), help=config_help("rsa_key_size"))
+
```
Comment (jsha) [inline certbot/cli.py]: Remove spurious diff

File: certbot-apache/certbot_apache/tests/testdata/debian_apache_2_4/multiple_vhosts/apache2/sites-available/ocsp-ssl.conf [hunk 13]
```diff
@@ -0,0 +1,36 @@
+<IfModule mod_ssl.c>
```
Comment (jsha) [inline certbot-apache/certbot_apache/tests/testdata/debian_apache_2_4/multiple_vhosts/apache2/sites-available/ocsp-ssl.conf]: You have two copies of this file - the one under letsencrypt-apache probably needs to go.

**Comment by pde** [comment]
@sagi should we give this back to @jsha for further review?

**Comment by sagi** [comment]
I don't think so. Everything I fixed there was based on his recommendations.

**Comment by jsha** [comment]
@sagi: Generally the code review culture is that after you fix things, the same reviewer goes back and checks that everything was fixed as specified. So I'll give this another round of review soon.

**Comment by jsha** [comment]
LGTM

**Pull request merged**